### DISCUSSION TITLE
Problem with spaces in path

### DISCUSSION THREAD

**Discussion by alfredoMitziC** [SchneiderInfosystems/FB4D#124] [Q&A] [answered]
Hi

I have this problem, when I try to put information in the database the following happens 

![imagen](https://user-images.githubusercontent.com/86743401/147169257-ee592195-0728-4a9e-9b9d-30dc8959748c.png)

Although I send the path with spaces, in the database it replaces it with plus signs and create a new node 

I have reviewed the examples but the same thing happens

Do you have any idea what may be happening or what I am doing wrong?

**Comment by SchneiderInfosystems** [comment, marked as answer]
The Realtime Database can handle spaces in the path but FB4D currently cannot.
This is related to the URL encoding in  `TFirebaseRequest.InternalSendRequestSynchronous`.
If you remove TURLEncoding.TEncodeOption.SpacesAsPlus in the constant Options in the following method in FB4D.Helpers in line 290 you can drive with spaces in the path too.

```
class function TFirebaseHelpers.EncodeResourceParams(
  Params: TRequestResourceParam): string;
const
  PathUnsafeChars: TURLEncoding.TUnsafeChars =
    [Ord('"'), Ord('<'), Ord('>'), Ord('^'), Ord('`'), Ord('{'), Ord('}'),
     Ord('|'), Ord('/'), Ord('\'), Ord('?'), Ord('#'), Ord('+')];
  Option: TURLEncoding.TEncodeOptions =
    [//TURLEncoding.TEncodeOption.SpacesAsPlus,
     TURLEncoding.TEncodeOption.EncodePercent];
var
  i: integer;
begin
  result := '';
  for i := low(Params) to high(Params) do
    result := result + '/' + TNetEncoding.URL.Encode(Params[i], PathUnsafeChars,
      Option);
end;
```

Because this is such a diversely used function, I am cautious about correcting it. Maybe other services will not run correctly anymore and spaces in paths are not used every day. If you can test this solution a bit and describe the tests, I can include the change definitive in an upcoming release. You would also have to test all other encoded characters in the PathUnsafeChars constant.

Thanks for this report.

**Reply by SchneiderInfosystems** [reply]
I decided not to make this change because the risk of further consequential problems seems too high compared to the benefit of having a Space correctly entered as an ID in the real-time DB. Note that you can use a Space in the ID and this works throughout FB4D for Put, Get, and the Listener. The Space character is simply converted to a Plus.
